### PR TITLE
[gbc c++] Generate copy and move constructors with allocator

### PR DESCRIPTION
**Pull request by ara-ayvazyan**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,6 +19,11 @@ different versioning scheme, following the Haskell community's
 * C# NuGet version: TBD
 * C# Comm NuGet version: TBD
 
+### `gbc` and Bond compiler library ###
+
+* C++ codegen now generates additional copy and move constructors with allocator
```
Comment (chwarr) [inline CHANGELOG.md]: Minor: consider "C++ codegen now generates copy and move constructors with an allocator argument when a custom allocator is used."

File: compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs [hunk 2]
```diff
@@ -276,6 +278,28 @@ namespace std
                 getAllocator BT_MetaFullName =  [lt|.get_allocator()|]
                 getAllocator _ = mempty
 
+        -- copy constructor with allocator
+        allocatorCopyCtor alloc = [lt|
```
Comment (chwarr) [inline compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs]: If we combine this into one function called something like `allocatorCopyMoveCtors`, we can share a ton of the identical code and just have `copyFieldInit` and `moveFieldInit` to handle the differences between wrapping with `std::move` and not.

Perhaps something like this will help too:

```
withMove x = [lt|std::move(#{x})|] -- and for not moving, just use id
```

File: compiler/tests/generated/alias_with_allocator_types.h [hunk 3]
```diff
@@ -46,6 +46,22 @@ namespace test
         foo(const foo&) = default;
 #endif
         
+        foo(const foo& other, const arena& allocator)
```
Comment (chwarr) [inline compiler/tests/generated/alias_with_allocator_types.h]: Should these constructors be templated so that they are only expanded if they are actually used?

File: compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs (hunk 2 above)
Comment (ara-ayvazyan) [reply compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs]: Updated.

File: CHANGELOG.md (hunk 1 above)
Comment (ara-ayvazyan) [reply CHANGELOG.md]: Updated.

File: compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs (hunk 2 above)
Comment (chwarr) [reply compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs]: The functions still look nearly identical to me. I was sort of expecting something more along the lines of this:

```
allocatorCtorBuilder argumentAdjuster = ...
allocatorCopyCtor = allocatorCtorBuilder id
allocatorMoveCtor = allocatorCtorBuilder withMove
    where withMove x = ...
```

Comment (ara-ayvazyan) [reply compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs]: I've not sent the updated commit yet :)

File: compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs [hunk 4]
```diff
@@ -278,17 +278,20 @@ namespace std
                 getAllocator BT_MetaFullName =  [lt|.get_allocator()|]
                 getAllocator _ = mempty
 
-        -- copy constructor with allocator
-        allocatorCopyCtor alloc = [lt|
-        #{declName}(const #{declName}&#{otherParam}, const #{alloc}&#{allocParam})#{initList}#{ctorBody}
-        |]
+        -- copy/move constructor with allocator
+        allocatorCopyOrMoveCtor fieldDecl paramValue alloc = [lt|
```
Comment (chwarr) [inline compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs]: `fieldDecl` make me think this is handling each field one-by-one. Perhaps `paramDecl` or `otherParamDecl`? (If `otherParamDecl` is picked, then consider `otherParamValue` instead of `paramValue`.)

**Comment by chwarr** [comment]
Please try to avoid rebasing during a review. It forces reviewers to re-review everything, including stuff they've already looked at, as GitHub can't handle tracking comments across rebases well.

File: compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs [hunk 5]
```diff
@@ -118,8 +119,8 @@ namespace std
     {
         #{newlineSepEnd 2 field structFields}#{defaultCtor}
 
-        #{copyCtor}
-        #{moveCtor}
+        #{copyCtor}#{optional allocatorCopyCtor allocator}
+        #{moveCtor}#{optional allocatorMoveCtor allocator}
```
Comment (chwarr) [inline compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs]: Consider moving the `optional` into `allocatorCopyOrMoveCtor` so we do the two checks at the same place.

File: cpp/inc/bond/core/maybe.h [hunk 6]
```diff
@@ -45,17 +53,23 @@ class maybe
...
     template <typename Allocator>
-    explicit maybe(Allocator& alloc)
+    explicit maybe(const Allocator& alloc)
         : _value(alloc),
           _nothing(true)
     {}
 
     template <typename Compare, typename Allocator>
-    explicit maybe(const Compare& comp, Allocator& alloc)
+    maybe(const Compare& comp, const Allocator& alloc)
```
Comment (chwarr) [inline cpp/inc/bond/core/maybe.h]: Consider checking where this is called to see whether we need such an overload at all.

Comment (ara-ayvazyan) [reply cpp/inc/bond/core/maybe.h]: Looks like we do not need this anymore since we moved to `C++11`. Will send a separate PR after I confirm it works on all supported compilers.

File: compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs (hunk 4 above)
Comment (ara-ayvazyan) [reply compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs]: Updated.

File: compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs (hunk 5 above)
Comment (ara-ayvazyan) [reply compiler/src/Language/Bond/Codegen/Cpp/Types_h.hs]: Updated.

File: compiler/Options.hs [hunk 7]
```diff
@@ -95,6 +96,7 @@ cpp = Cpp
     , core_enabled = True &= explicit &= name "core" &= help "Generate core serialization definitions (true by default, --core=false to disable)"
     , comm_enabled = False &= explicit &= name "comm" &= help "Generate comm definitions"
     , grpc_enabled = False &= explicit &= name "grpc" &= help "Generate gRPC definitions"
+    , alloc_ctors_enabled = False &= explicit &= name "alloc_ctors" &= help "Generate constructors with allocator argument"
```
Comment (ara-ayvazyan) [inline compiler/Options.hs]: Should be `alloc-ctors` (with `-`).

File: compiler/src/Language/Bond/Codegen/Java/Class_java.hs [hunk 8]
```diff
@@ -235,13 +235,16 @@ makeStructBondTypeMember_StructBondTypeBuilderImpl declName declParams = [lt|
...
-                return new StructBondTypeImpl(#{ifThenElse (null declParams) "null" "new org.bondlib.GenericTypeSpecialization(genericTypeArguments)"});
+                return new StructBondTypeImpl(#{ifThenElse (null declParams) nullStr "new org.bondlib.GenericTypeSpecialization(genericTypeArguments)"});
             }
 
             static void register() {
                 registerStructType(#{declName}.class, new StructBondTypeBuilderImpl());
             }
         }|]
+    where
+        nullStr :: Text
```
Comment (ara-ayvazyan) [inline compiler/src/Language/Bond/Codegen/Java/Class_java.hs]: `nullText`

Comment (chwarr) [reply compiler/src/Language/Bond/Codegen/Java/Class_java.hs]: You should be able to shorten this to one line:

```
nullText = "null" :: Text
```

File: compiler/src/Language/Bond/Util.hs [hunk 9]
```diff
@@ -70,6 +71,11 @@ sepBy s f (x:xs)
 optional :: (Monoid m) => (a -> m) -> Maybe a -> m
 optional = maybe mempty
 
+-- takes a predicate and two items and outputs one based on predicate
```
Comment (chwarr) [inline compiler/src/Language/Bond/Util.hs]: Minor: this doesn't take a predicate. That would be more like `(b -> Bool) -> a -> a`.
Maybe just "if/then/else as a function"?

Comment (ara-ayvazyan) [reply compiler/src/Language/Bond/Util.hs]: Updated to: `if-then-else as a function`
